pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -0,0 +1,199 @@
+
+// tomlURL is an abstraction required to unmarshal the non-primitive and
+// non-local url.URL type when loading the toml config.
+type tomlURL struct {
+	url url.URL
+}
+
+// Mirror represents a mirror. Mirrors can be used as pull-through caches for
+// registries.
+type Mirror struct {
+	// The mirror's URL
+	URL tomlURL `toml:"url"`
+	// If true, certs verification will be skipped
+	Insecure bool `toml:"insecure"`
+}
+
+// Registry represents a registry.
+type Registry struct {
+	// The registry's URL
+	URL tomlURL `toml:"url"`
+	// If true, pulling from the registry will be blocked
+	Blocked bool `toml:"blocked"`
+	// If true, certs verification will be skipped
+	Insecure bool `toml:"insecure"`
+	// If true, the registry can be used when pulling an unqualified image
+	Search bool `toml:"unqualified-search"`
+	// If it prefixes an image, the registry will be used for pulling
+	// (defaults to the URL without the URI scheme)
+	Prefix string `toml:"prefix"`
+	// Pull through mirrors (cannot further be mirrored)
+	Mirrors []Mirror `toml:"mirror"`
+}
+
+// tomlConfig is the data type used to unmarshal the toml config.
+type tomlConfig struct {
+	Registries []Registry `toml:"registry"`
+}
+
+// UnmashalText interprets and parses text as a net.URL type and assigns it to
+// the tomlURL's url field.
+func (r *tomlURL) UnmarshalText(text []byte) (err error) {
+	r.url, err = parseURL(string(text))
+	return err
+}
+
+// parseURL parses the input string, performs some sanity checks and returns
+// a url.URL.  The input must be a valid URI with an "http" or "https" scheme,
+// a specified host and an empty URI user.  Otherwise, an error is returned.
+func parseURL(input string) (url.URL, error) {
+	uri, err := url.Parse(input)
+	if err != nil {
+		return url.URL{}, fmt.Errorf("error parsing URL %s: %v", input, err)
+	}
+
+	// only https and http are valid URI schemes
+	if uri.Scheme == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI scheme: %s", input)
+	}
+	if uri.Scheme != "https" && uri.Scheme != "http" {
+		return url.URL{}, fmt.Errorf("unsupported URI scheme: %s", input)
+	}
+
+	// a host must be specified
+	if uri.Host == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI host: %s", input)
+	}
+
+	// user must be empty
+	if uri.User != nil {
+		// strip password for security reasons
+		uri.User = url.UserPassword(uri.User.Username(), "xxxxxx")
+		return url.URL{}, fmt.Errorf("unsupported username/password: %q", uri)
+	}
+
+	return *uri, nil
+}
+
+// GetRegistries loads and returns the registries specified in the config.
+func GetRegistries(ctx *types.SystemContext) ([]Registry, error) {
+	config, err := loadRegistryConf(ctx)
+	if err != nil {
+		return []Registry{}, err
+	}
+	// if no prefix is specified default to specified URL without the URI scheme
+	for i, reg := range config.Registries {
+		if reg.Prefix == "" {
+			prefix := strings.TrimPrefix(reg.URL.url.String(), reg.URL.url.Scheme)
```

Comment:
Doesn’t this strip only `http` from `http://example.com`, leaving `prefix` set to `://example.com`?

Reply:
Good catch, I will check this. Once we have an agreement on the design, I will make sure to add unit tests.